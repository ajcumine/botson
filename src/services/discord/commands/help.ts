import { Message } from 'discord.js';
import { PREFIX } from '../';
import logger from '../../logger';
import commands, { Command } from './';

export const execute = async (message: Message, args: string[]) => {
  const data: string[] = [];
  if (!args.length) {
    data.push("Here's a list of all my commands:");
    data.push(
      Object.keys(commands)
        .map(commandKey => `\`${PREFIX}${commandKey}\``)
        .join(`, `),
    );
    data.push(
      `\nYou can send \`${PREFIX}help [command name]\` to get info on a specific command!`,
    );

    try {
      await message.author.send(data, { split: true });
      if (message.channel.type === 'dm') return;
      return message.reply("I've sent you a DM with all my commands!");
    } catch (error) {
      logger.error(`Could not send help DM to ${message.author.tag}.\n`, error);
      message.reply("it seems like I can't DM you! Do you have DMs disabled?");
    }
  }

  const commandName = args[0];
  const command = commands[commandName];

  if (!command) {
    return message.reply("that's not a valid command!");
  }

  data.push(`**Name:** ${commandName}`);
  data.push(`**Description:** ${command.description}`);
  if (command.usage) {
    data.push(`**Usage:** \`${PREFIX}${commandName} ${command.usage}\``);
  }

  return message.channel.send(data, { split: true });
};

const help: Command = {
  execute,
  description:
    'Provide a list of all the bot commands or specific info for a single command',
  argsRequired: false,
  guildOnly: false,
  usage: '[command name]',
};

export default help;
