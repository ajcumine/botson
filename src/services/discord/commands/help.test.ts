import help, { execute } from './help';

describe('!help command', () => {
  it('handles !help commands', () => {
    expect(help).toMatchSnapshot();
  });

  describe('execute', () => {
    describe('if there are no args', async () => {
      it('replies to the author in channel and DMs help commands', async () => {
        let sentMessageData;
        const authorSendSpy = jest.fn(data => {
          sentMessageData = data;
          return Promise.resolve();
        });
        const replySpy = jest.fn();
        const mockMessage = {
          author: {
            send: authorSendSpy,
          },
          channel: {
            type: 'notDm',
          },
          reply: replySpy,
        };
        await execute(mockMessage, []);
        expect.assertions(3);
        expect(authorSendSpy).toHaveBeenCalled();
        expect(sentMessageData).toMatchSnapshot();
        expect(replySpy).toHaveBeenCalledWith(
          "I've sent you a DM with all my commands!",
        );
      });
    });

    describe('if there are args', () => {
      describe('if the arg is an invalid command', () => {
        it('informs the author', () => {
          const replySpy = jest.fn();
          const mockMessage = {
            reply: replySpy,
          };
          execute(mockMessage, ['_']);
          expect(replySpy).toHaveBeenCalledWith("that's not a valid command!");
        });
      });

      describe('if the arg is a valid command', () => {
        it('replies to the author in channel with command instructions', () => {
          let sentMessageData;
          const sendSpy = jest.fn(data => (sentMessageData = data));
          const mockMessage = {
            channel: {
              send: sendSpy,
            },
          };
          execute(mockMessage, ['ping']);
          expect(sentMessageData).toMatchSnapshot();
          expect(sendSpy).toHaveBeenCalled();
        });
      });
    });
  });
});
