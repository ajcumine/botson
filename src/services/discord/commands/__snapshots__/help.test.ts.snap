// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`!help command execute if there are args if the arg is a valid command replies to the author in channel with command instructions 1`] = `
Array [
  "**Name:** ping",
  "**Description:** Ping!",
]
`;

exports[`!help command execute if there are no args replies to the author in channel and DMs help commands 1`] = `
Array [
  "Here's a list of all my commands:",
  "\`!help\`, \`!ping\`, \`!prune\`, \`!role\`",
  "
You can send \`!help [command name]\` to get info on a specific command!",
]
`;

exports[`!help command handles !help commands 1`] = `
Object {
  "argsRequired": false,
  "description": "Provide a list of all the bot commands or specific info for a single command",
  "execute": [Function],
  "guildOnly": false,
  "usage": "[command name]",
}
`;
